lib/puppet/pops/loader/loader_paths.rb

```diff
@@ -221,6 +226,20 @@ def relative_path
       TASKS_PATH
     end
 
+    def typed_name(type, name_authority, relative_path, module_name)
+      n = ''
+      n << module_name unless module_name.nil?
+
+      # Remove extension regardless of what it is. A task name cannot contain dots
+      relative_path = relative_path.sub(/\.[^\/]*\z/, '')
```

Comment:
We use `File.basename(relative_path, '.*')` elsewhere. It only returns the file component, not the parent, so you'd need to do something like `File.join(File.dirname(relative_path), File.basename(relative_path, '.*'))`

Reply:
@joshcooper , I'd happily change this to what you propose if you have strong feelings about that. But claiming that it's needed isn't right. Two things are well known at this point. There is no backslash in the path and dot's are not allowed in a task name. Hence, it's perfectly safe to strip the dot and all characters that are not slashes until the end of string is reached and doing it with one regexp it's probably a lot faster than what you suggest.

Reply:
Also, please note that the whole `SmartPath` concept is built with linux-style paths in mind. There are several places that needs to change, should we decide to change that. The change isn't needed though, because it all acts on paths produced by Dir.glob, cached, sometimes mangled by Pathname and then cached.

Reply:
> whole SmartPath concept is built with linux-style paths in mind... because it all acts on paths produced by Dir.glob, cached, and sometimes mangled by Pathname

Pathname is not guaranteed to return forward slashes. On Windows the following

```
require 'pathname'
require 'win32-dir'
puts Pathname.new(Dir.pwd)
```

produces `C:\cygwin64\home\Administrator`

Reply:
That doesn't matter. We never feed backslash paths to Pathname. All paths here stem from Dir.glob and Dir.glob never returns paths with backslash. Everything depends on that.

Reply:
> All paths here stem from Dir.glob and Dir.glob never returns paths with backslash.

Earlier you said the relative paths from Dir.glob are sometimes mangled by Pathname. If you ever try to expand the resulting Pathname using the current working directory as a starting point, you may get backslashes on Windows. Maybe that will never happen.

> doing it with one regexp it's probably a lot faster than what you suggest

Maybe... but it seems like a premature optimization given `File.basename` is implemented in native code: https://github.com/ruby/ruby/blob/ruby_2_4/file.c#L4114

Reply:
Yes, paths obtained from `Dir.glob` are mangled with `Pathname`. But `Pathname` doesn't change a '/' path to '\'.

What you suggest isn't limited to `File.basename`. It also involves a `File.join`, and a `File.dirname`. Three calls compare to one `Regexp.sub` (also native). Further more `File.basename('foo.tar.gz', '.*')` returns 'foo.tar'. That's not what I want. I want 'foo'. A Task cannot have dots in the name.

Reply:
I'm curious about the `win32-dir` and `Dir.glob` that you refer to.

When I use `Dir.pwd` on Windows I get forward slashes and there's no need to do `require win32-dir`. Perhaps that's ruby version specific?

Reply:
The win32-dir gem modifies ruby's behavior to return a more "Windows friendly" path. For various reasons, we haven't been able to drop that dependency yet.

Reply:
See https://github.com/djberg96/win32-dir/blob/d45d400d7db2d578f078ccbde7a8fc5096857525/lib/win32/dir.rb#L182